### PR TITLE
WIP: Event onClientMTAFocusChange

### PR DESCRIPTION
**Pull request by CrosRoad95**
This is event similar to `onClientMinimize` but, insted of checking did mta windows was minimized, this events are triggered when you lost focus on mta window.
If someone play on rpg server in window mode, then he never minimize them. This event can help programmer handle when player is afk

Example script clientside
```lua
addEventHandler("onClientMTAFocusChange", root, function (focus)
  outputChatBox(inspect(focus))
end)
```
Print `false` when you lost focus on mta window, when you back to game, will print `true`

### CODE REVIEW

File: Client/mods/deathmatch/logic/CClientGame.h [hunk 1]
```diff
@@ -740,6 +740,7 @@ class CClientGame
     bool                                m_bBeingDeleted;        // To enable speedy disconnect
 
     bool                                m_bWasMinimized;
+    bool                                m_bWasFocus;
```
Comment (Citizen01) [inline Client/mods/deathmatch/logic/CClientGame.h]: `m_bWasFocused` would be better

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 2]
```diff
@@ -3959,6 +3961,24 @@ void CClientGame::IdleHandler ( void )
                 m_pManager->GetSoundManager ()->SetMinimizeMuted ( true );
         }
     }
+    if ( g_pCore->IsFocused() )
+    {
+        if (!m_bWasMinimized)
```
Comment (Citizen01) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: You probably wanted to use `m_bWasFocus` instead (or `m_bWasFocused` if you rename it)

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 3]
```diff
@@ -3959,6 +3961,24 @@ void CClientGame::IdleHandler ( void )
...
+        {
+            m_bWasMinimized = true;
```
Comment (Citizen01) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: Same: `m_bWasFocus` instead (or `m_bWasFocused` if you rename it)

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 4]
```diff
@@ -3959,6 +3961,24 @@ void CClientGame::IdleHandler ( void )
...
+            CLuaArguments Arguments;
+            m_pRootEntity->CallEvent("onClientMTAFocus", Arguments, false);
+        }
+    }
+    else
+    {
+        if (m_bWasMinimized)
```
Comment (Citizen01) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: Same: `m_bWasFocus` instead (or `m_bWasFocused` if you rename it)

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 5]
```diff
@@ -3959,6 +3961,24 @@ void CClientGame::IdleHandler ( void )
...
+        {
+            m_bWasMinimized = false;
```
Comment (Citizen01) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: Same: `m_bWasFocus` instead (or `m_bWasFocused` if you rename it)

**Review by Citizen01 — COMMENTED** [review]
Thanks for the work you put into MTA lately.
Here are my comments and also in JS the name of such events are `Focus` and `Blur` but the latest doesn't sound familiar for someone looking up into the wiki ... But there is also `FocusIn` and `FocusOut`.
You event names are perfectly fine to me, just throwing in some variants in case someone wants to debate over it. (Do not rename it until a decision is made).

**Comment by Citizen01** [comment]
Also, I'm concerned about the performance of `CCore::IsFocused`:
```cpp
 bool IsFocused ( void ) { return ( GetForegroundWindow ( ) == GetHookedWindow ( ) ); };
```
:warning: **I'm asking to whoever knows:** Is calling that method every frame okay ?

I really feel like we need some sort of optimization like it has been done with `IsWindowMinimized` (updated once every 50 ms if I read the code correctly from Github).

**Comment by botder** [comment]
I think it would be much better to wait for the WndProc events, then delegate the event from Core to Deathmatch and finally trigger the event.

**Comment by botder** [comment]
I assume you have to add code to these following lines, which handle the focus loss/gain of the GTA window:
https://github.com/multitheftauto/mtasa-blue/blob/882f57294bce0c795080050a9a6ffa234c7bd75a/Client/core/CMessageLoopHook.cpp#L127-L134

From there you can call into the deathmatch module to handle the focus change.

**Comment by botder** [comment]
Mantis issue: https://bugs.mtasa.com/view.php?id=8651

**Comment by qaisjp** [comment]
stale

**Pull request closed without merging**